VFB-407 Update dietary requirements item lists

mihaicutaiulia opened the pull request:
Jira ticket - [VFB-407](https://softwiretech.atlassian.net/jira/software/c/projects/VFB/boards/1130?selectedIssue=VFB-407)

## What's changed
- Introduced **edit modal** for dietary requirements:
  - Added a new `EditDietaryRequirementsModal` React component for editing dietary requirements using a modal dialog
  - Users can now select included/excluded items for each dietary requirement type (halal, vegetarian, vegan, etc.) and save their changes.
- Added an `Edit Dietary Requirements` button to open the modal.
  - Integrated the new modal with the main `DietaryRequirementsTable` component.
  - Improved user feedback with success and warning alerts in the modal.
- Modified the Supabase table `dietary_requirements` to set a default value of `not_specified` for all dietary status columns.
- Added a migration to include the `dietary_requirements` table in **Supabase realtime updates**.

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| <img width="1847" height="762" alt="image" src="https://github.com/user-attachments/assets/31dfea23-c9d7-465a-b678-29c2ad78b7ae" /> | <img width="1844" height="795" alt="image" src="https://github.com/user-attachments/assets/c550d210-0d23-4abe-af6f-108d499a9356" /> |
| | <img width="573" height="952" alt="image" src="https://github.com/user-attachments/assets/0f0e066a-12e4-4434-a337-696690599cbe" /> |

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [ ] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test`

If you have made any changes to the database...
  - [x] The migration files are up-to-date with my final set up (`npx supabase db diff -f <name_of_migration>` should create nothing at this point)
  - [x] I have updated the typescript definitions for the database with `npm run generate_types:local`
  - [x] I have modified the seed in `supabase/seed/seed.ts` if appropriate
  - [x] If I have modified the seed, I have also generated the seed with `npm run db:generate_seed` 
  - [x] With my final set up, I can run `npm run reset_supabase` without any errors.
  - [ ] (Just before merging the ticket) I have updated the timestamps of the new migration files so that they are after all existing migration files.
  - [ ] I have taken reasonable measures to ensure constraint violations do not happen when the migration occurs, e.g.:
    - audit_log FK constraints that could be violated
    - any constraints that I have added, e.g. `set not null`
  - [ ] I have checked that migration can happen successfully without resetting the database, doing the following.
    - Make sure you have rebased your branch onto dev or merged dev into your branch.
    - Checkout the dev branch
    - Run `npm run post_checkout` to reset the database, including the seed data.
    - Checkout your branch
    - Run `npx supabase migrations up --include-all --local` to run all outstanding migration files
    - Check that the resulting database is what you expect, bar any seed data changes.

stefanghenescu-work committed:
VFB-407: add table for dietary requirements in db

stefanghenescu-work committed:
VFB-407: table for dietary is shown but needs work with pagination

mihaicutaiulia committed:
VFB-407: generate database types

stefanghenescu-work committed:
VFB-407: add edit button and model for selecting items for every dietary requirement.
Edits are also visible in the DB.

stefanghenescu-work committed:
VFB-407: make form look better and cleaner.
resolve linter errors

stefanghenescu-work committed:
VFB-407: form uses Model component now

mihaicutaiulia committed:
VFB-407: set realtime on for dietary_requirements table

mihaicutaiulia committed:
VFB-407: generate seed for dietary_requirements

anamariasuditu commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,239 @@
+import Modal from "@/components/Modal/Modal";
+
+interface Props {
+    open: boolean;
```
maybe you could rename the variable to `isOpen`

claudia-golaes-dev commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,239 @@
+                        case "pet_food":
+                            value = row.pet_food;
+                            break;
+                    }
```
The switch function can be replaced with `value = row[selectedType as keyof typeof row] || null;`

mihaicutaiulia replied:
i replaced it. Thank you! 🫶

LeonardCiobanu-Softwire commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,239 @@
+            const items = await fetchAllItems();
+            setItems(items);
+
+            const { data } = await supabase.from("dietary_requirements_plus").select();
```
Could you add the error field in case the query is unsuccessful?

LeonardCiobanu-Softwire reviewed (commented):
👍

mihaicutaiulia committed:
VFB-407: fix migrations

mihaicutaiulia committed:
VFB-407: solved comments and fixed seed after rebase

VFB-407: fixed seed after rebase

VFB-407: remove redundant variable initializer

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsActions.ts:
```diff
@@ -64,6 +76,24 @@ export const fetchDietaryRequirementsForTable =
+    const { data, error } = await supabase.from("lists").select("primary_key, item_name");
+
+    if (error) {
+        throw error;
```
We shouldn't throw the error here

AlexandraIrinaMardale replied:
We should either make sure we handle it upstream, or handle it here

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+    onClose: () => void;
+}
+
+export const EditDietaryRequirementsModal: React.FC<Props> = ({ isOpen, onClose }) => {
```
There is a warning on `onClose`, can you please fix it?

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+            setExcluded(excludedSet);
+        };
+
+        fetchData();
```
There's a warning here as well

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+            const excludedSet = new Set<string>();
+
+            items.forEach((item) => {
+                const row = data.find((iterRow) => iterRow.id === item.id);
```
items and data contain the exact same id's, so you don't need this

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+                const row = data.find((iterRow) => iterRow.id === item.id);
+
+                if (row) {
+                    const value = row[selectedType as keyof typeof row] || null;
```
Do you need `|| null`?

mihaicutaiulia replied:
no

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+
+            setIncluded(includedSet);
+            setExcluded(excludedSet);
+        };
```
And instead of doing this you can set the included and excluded items using the [spread operator](https://www.w3schools.com/react/react_es6_spread.asp)

AlexandraIrinaMardale commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,211 @@
+            }
+
+            const includedSet = new Set<string>();
+            const excludedSet = new Set<string>();
```
Can you use simple arrays instead of sets? There isn't really any risk of duplicated entries here

mihaicutaiulia committed:
VFB-407: refactor logic for database fetch and update

mihaicutaiulia committed:
VFB-407: update migration to have default values for dietary requirements

mihaicutaiulia committed:
VFB-407: sort items by item_name and used the spread operator for included/excluded arrays

mihaicutaiulia committed:
VFB-407: refactor types

stefanghenescu-work committed:
VFB-407: remove unnecessary types

stefanghenescu-work committed:
VFB-407: add warnings when saving dietary requirements

stefanghenescu-work committed:
VFB-407: resolve warnings when saving dietary requirements for edge cases

stefanghenescu-work committed:
VFB-407: change migration for field type

DanielaNegrut commented:
I see that Alexandra already left some review comments, could you please add a react emoji to those that are solved @mihaicutaiulia?

mihaicutaiulia commented:
> I see that Alexandra already left some review comments, could you please add a react emoji to those that are solved @mihaicutaiulia?

I’ll go through her comments and add a reaction to the ones that are now resolved. Since Alexandra’s review, however, we’ve refactored most of the implementation to avoid unnecessary updates to the database.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+interface Props {
+    isOpen: boolean;
+    onClose: () => void;
+}
```
This seems like a simple and re-usable prop interface. Is there any place we might also use it? If not, I am happy to leave it in this file.

mihaicutaiulia replied:
I can't think of any now

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+            .order("item_name");
+
+        if (error) {
+            console.error("Error fetching dietary requirements data:", error);
```
Is this a usual practice in the project? Logging the error in browser?

mihaicutaiulia replied:
This behavior will be changed in #743 to send the logs using `audit log` instead of displaying them in the browser console.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+            const isNowIncluded = newIncluded.includes(item.id);
+            const isNowExcluded = newExcluded.includes(item.id);
+
+            return wasIncluded !== isNowIncluded || wasExcluded !== isNowExcluded;
```
Let's create a helper function for these so we will reduce the size of this handler a bit.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+        setHasChanges(false);
+        setWarningSaveMessage("");
+
+        void fetchData(selectedType);
```
I'm curious why we are refetching the data

stefanghenescu-work replied:
To ensure the `Alert` component correctly detects changes, we need to update the `initialIncluded` and `initialExcluded` arrays. Since these arrays are set in `fetchData`, they should be refreshed whenever the table is updated which is done after submit.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+            }
+        }
+        return true;
+    }
```
This seems like a utility function.

mihaicutaiulia replied:
Yes. I moved the function outside of `EditDietaryRequirementsModal`

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+
+export const EditDietaryRequirementsModal: React.FC<Props> = ({ isOpen, onClose }) => {
+    const [items, setItems] = useState<DietaryRequirementsPlusTableRow[]>([]);
+    const [selectedType, setSelectedType] = useState<string>("halal");
```
Can't these be of type `keyof BaseDietaryRequirements` instead of string? So we won't do so many casts.

mihaicutaiulia replied:
👍 The `BaseDietaryRequirements` is a new addition, thank you for pointing this out

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+            return false;
+        }
+        firstArray.sort();
+        secondArray.sort();
```
Isn't this going to modify the arrays we pass?

stefanghenescu-work replied:
We will save a copy of the arrays to prevent this issue.
Up to now, the order of elements hasn’t mattered, since we only check whether an item exists in the array, not its position.

mihaicutaiulia replied:
In the updated implementation, we make copies of the arrays and sort those copies, so the original arrays remain unchanged.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+        return true;
+    }
+
+    const handleToggle = (id: string | null, type: "included" | "excluded"): void => {
```
I think that this function is a bit confusing 🤔 

Is there anything we can do so that it is clearer what happens?

stefanghenescu-work replied:
Yes, I will refactor it!

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsTable.tsx:
```diff
@@ -37,7 +41,7 @@ const DietaryRequirementsTable: React.FC = () => {
             .on(
                 "postgres_changes",
-                { event: "*", schema: "public", table: "dietary_requirements_plus" },
+                { event: "*", schema: "public", table: "dietary_requirements" },
```
I'm curious about this renaming. I think there is some context that I am missing.

mihaicutaiulia replied:
`dietary_requirements_plus` is actually a view based on the `dietary_requirements` table.

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsForm.tsx:
```diff
@@ -0,0 +1,305 @@
+    id: string;
+};
+
+export const EditDietaryRequirementsModal: React.FC<Props> = ({ isOpen, onClose }) => {
```
I see that this file is named something different then the component it exports. Is this a basic practice in the codebase?

mihaicutaiulia replied:
I don't think so. I will rename the file

DanielaNegrut commented on src/app/admin/dieteryRequirementsTable/DietaryRequirementsTable.tsx:
```diff
@@ -113,6 +126,9 @@ const DietaryRequirementsTable: React.FC = () => {
                     aria-label="Dietary Requirements Table"
                     columns={dietaryRequirementsColumns}
                     getRowHeight={() => "auto"}
+                    editMode="row"
+                    rowModesModel={rowModesModel}
+                    onRowModesModelChange={setRowModesModel}
```
What do these changes do?

mihaicutaiulia replied:
The table was initially read-only; these changes enable editing functionality.

DanielaNegrut commented on supabase/migrations/20250728083754_add_dietary_requirements_table.sql:
I see that we are modifying an existing migration. What are the implications of that?

mihaicutaiulia replied:
Since this migration hasn’t been run on `intern-dev` yet (it was introduced in #730), modifying it doesn’t pose any issues.

mihaicutaiulia committed:
VFB-407: solve comments

mihaicutaiulia committed:
VFB-407: refactor checkArraysAreEqual

stefanghenescu-work committed:
VFB-407: resolve unsave changes and refactor `handleToggle` function

mihaicutaiulia committed:
VFB-429 Audit functionality for Dietary Requirements table (#743)

mihaicutaiulia committed:
VFB-408 Smart Shopping Lists (#741)

Co-authored-by: Stefan Ghenescu <stefan.ghenescu@softwire.com>

stefanghenescu-work committed:
VFB-407: table should not close after save
